### PR TITLE
Task 1/vulkan basic initialization

### PR DESCRIPTION
**Pull request by krupitskas**

### CODE REVIEW

File: VorpalEngine/CMakeLists.txt [hunk 1]
```diff
@@ -50,7 +50,14 @@ include(loguru)
 ## STB
 include(stb)
 
-find_library(Vulkan REQUIRED)
+
+if (UNIX)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
```
Comment (MrDmitry) [inline VorpalEngine/CMakeLists.txt]: This is done on line 21

File: VorpalEngine/CMakeLists.txt [hunk 2]
```diff
@@ -101,3 +108,6 @@ add_library(${VORPAL_ENGINE_LIB}
 )
 
 target_link_libraries(${VORPAL_ENGINE_LIB} ${GLFW_LIB} ${VULKAN_LIBRARY})
+if (UNIX)
+    target_link_libraries(${VORPAL_ENGINE_LIB} ${VULKAN_LIBRARY})
```
Comment (MrDmitry) [inline VorpalEngine/CMakeLists.txt]: This is done on line 110

File: VorpalEngine/data/shaders/shader.frag [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#version 450
+#extension GL_ARB_separate_shader_objects : enable
+
+layout(location = 0) in vec3 fragColor;
+
+layout(location = 0) out vec4 outColor;
+
+void main() {
+    outColor = vec4(fragColor, 1.0);
+}
```
Comment (MrDmitry) [inline VorpalEngine/data/shaders/shader.frag]: Please add empty line at the end of file

File: VorpalEngine/data/shaders/shader.vert [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+vec3 colors[3] = vec3[](
+    vec3(1.0, 0.0, 0.0),
+    vec3(0.0, 1.0, 0.0),
+    vec3(0.0, 0.0, 1.0)
+);
+
+void main() {
+    gl_Position = vec4(positions[gl_VertexIndex], 0.0, 1.0);
+    fragColor = colors[gl_VertexIndex];
+}
```
Comment (MrDmitry) [inline VorpalEngine/data/shaders/shader.vert]: Please add empty line

File: VorpalEngine/include/vorpal/video/Renderer.hpp [hunk 5]
```diff
@@ -13,6 +13,24 @@ namespace vp
 {
     namespace video
     {
+        struct QueueFamilyIndices
+        {
+            int graphicsFamily = -1;
+            int presentFamily = -1;
+
+            bool isComplete()
```
Comment (MrDmitry) [inline VorpalEngine/include/vorpal/video/Renderer.hpp]: Can also be const

File: VorpalEngine/sources/video/Renderer.cpp [hunk 6]
```diff
@@ -1,40 +1,85 @@
...
-
 #include <vorpal/utility/Logger.hpp>
+#include <vorpal/utility/asset/SimpleStorage.hpp>
 
 #include <cstring>
+#include <iostream>
+#include <set>
+#include <algorithm>
+
+//TODO temp until Dmitry will fix load code
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: could you use doxygen `@todo` tag?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 7]
```diff
@@ -1,40 +1,85 @@
...
 {
 namespace video
 {
 
-VkResult CreateDebugReportCallbackEXT(VkInstance instance,
-    const VkDebugReportCallbackCreateInfoEXT* pCreateInfo,
-    const VkAllocationCallbacks* pAllocator,
-    VkDebugReportCallbackEXT* pCallback)
+static VKAPI_ATTR VkBool32 VKAPI_CALL DebugCallback(VkDebugReportFlagsEXT flags,
+                                                    VkDebugReportObjectTypeEXT objType,
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: I don't think we should align to the right (to the opening bracket) but instead use 1 extra indent on the next lines

File: VorpalEngine/sources/video/Renderer.cpp [hunk 8]
```diff
@@ -1,40 +1,85 @@
...
 Renderer::Renderer()
     : m_isInitialized(false)
     , m_pWindow(nullptr)
     , m_vkInstance(VK_NULL_HANDLE)
+    , m_vkPhysicalDevice(VK_NULL_HANDLE)
     , m_validationLayers( {"VK_LAYER_LUNARG_standard_validation"} )
+    , m_deviceExtensions( {VK_KHR_SWAPCHAIN_EXTENSION_NAME})
 {
-
+    m_vkSwapChain = VK_NULL_HANDLE;
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp:62]: Why part is initialized in init list and part in constructor itself?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 9]
```diff
@@ -86,41 +140,252 @@ bool Renderer::Init()
...
+
+    std::vector<VkQueueFamilyProperties> queueFamilies(queueFamilyCount);
+    vkGetPhysicalDeviceQueueFamilyProperties(device, &queueFamilyCount, queueFamilies.data());
+
+    int i = 0;
+    for (const auto& queueFamily : queueFamilies) {
+        if (queueFamily.queueCount > 0 && queueFamily.queueFlags & VK_QUEUE_GRAPHICS_BIT) {
+            indices.graphicsFamily = i;
+        }
+        //TODO optimize?
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: `@todo`

File: VorpalEngine/sources/video/Renderer.cpp [hunk 10]
```diff
@@ -86,41 +140,252 @@ bool Renderer::Init()
...
+        vkGetPhysicalDeviceSurfaceSupportKHR(device, i, m_vkWindowSurface, &presentSupport);
+
+        if (queueFamily.queueCount > 0 && presentSupport) {
+             indices.presentFamily = i;
+         }
+
+        if (indices.isComplete()) {
+            break;
+        }
+        i++;
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: `++i`?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 11]
```diff
@@ -86,41 +140,252 @@ bool Renderer::Init()
...
+        details.presentModes.resize(presentModeCount);
+        vkGetPhysicalDeviceSurfacePresentModesKHR(device, m_vkWindowSurface, &presentModeCount, details.presentModes.data());
+    }
+    return details;
+}
+
+VkSurfaceFormatKHR Renderer::ChooseSwapSurfaceFormat(const std::vector<VkSurfaceFormatKHR> &availableFormats)
+{
+    if (availableFormats.size() == 1 && availableFormats[0].format == VK_FORMAT_UNDEFINED) {
+           return {VK_FORMAT_B8G8R8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR};
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: indents?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 12]
```diff
@@ -86,41 +140,252 @@ bool Renderer::Init()
...
+           }
+       }
+
+       return availableFormats[0];
+}
+
+VkPresentModeKHR Renderer::ChooseSwapPresentMode(const std::vector<VkPresentModeKHR> availablePresentModes)
+{
+    for (const auto& availablePresentMode : availablePresentModes) {
+           if (availablePresentMode == VK_PRESENT_MODE_MAILBOX_KHR) {
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: indents?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 13]
```diff
@@ -164,6 +429,597 @@ bool Renderer::CreateInstance()
...
+
+    VkRenderPassCreateInfo renderPassInfo = {};
+    renderPassInfo.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
+    renderPassInfo.attachmentCount = 1;
+    renderPassInfo.pAttachments = &colorAttachment;
+    renderPassInfo.subpassCount = 1;
+    renderPassInfo.pSubpasses = &subpass;
+
+    if (vkCreateRenderPass(m_vkLogicalDevice, &renderPassInfo, nullptr, &m_renderPass) != VK_SUCCESS) {
+       LOG_ERROR("Failed to create render pass!");
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp:622]: 7 spaces indent? :-1:

File: VorpalEngine/sources/video/Renderer.cpp [hunk 14]
```diff
@@ -164,6 +429,597 @@ bool Renderer::CreateInstance()
...
+    //            || !simpleFragShaderHandler.IsValid()) {
+    //        LOG_ERROR("Can't find shaders!");
+    //        return false;
+    //    }
+
+    auto vertShaderCode = readFile("data/shaders/vert.spv");
+    auto fragShaderCode = readFile("data/shaders/frag.spv");
+    VkShaderModule vertShaderModule = {};
+    VkShaderModule fragShaderModule = {};
+    bool shadersCreated = !CreateShaderModule(vertShaderCode, vertShaderModule)
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: variable is named `shadersCreated` but it holds invert of that statement. Rename or change logic?

File: VorpalEngine/sources/video/Renderer.cpp [hunk 15]
```diff
@@ -164,6 +429,597 @@ bool Renderer::CreateInstance()
...
+    }
+    if(deviceProperties.deviceType == VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU
+            && indices.isComplete()
+            && extensionsSupported
+            && swapChainAcceptable) {
+        LOG_INFO("Picked as main GPU : %s", deviceProperties.deviceName);
+        m_gpuName = deviceProperties.deviceName;
+        return true;
+    } else
+        return false;
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp:929]: please don't omit curly braces ever

File: VorpalEngine/sources/video/Renderer.cpp [hunk 16]
```diff
@@ -202,51 +1058,37 @@ bool Renderer::CheckValidationLayerSupport() const
...
 
 std::vector<const char*> Renderer::GetRequiredExtensions()
-{
-    std::vector<const char*> extensions;
-
+{    
+    std::vector<const char*> m_extensions;
     unsigned int glfwExtensionCount = 0;
     const char** glfwExtensions;
     glfwExtensions = glfwGetRequiredInstanceExtensions(&glfwExtensionCount);
```
Comment (MrDmitry) [inline VorpalEngine/sources/video/Renderer.cpp]: could be done in one line with declaration

**Review by MrDmitry — DISMISSED** [review]
Refer to in-file comments

File: VorpalEngine/sources/video/Renderer.cpp (hunk 13 above)
Comment (krupitskas) [reply VorpalEngine/sources/video/Renderer.cpp:622]: Nope, 8 here. I just counted in notepad++

**Pull request merged**